packages/ai-jsx/src/core/memoize.tsx

```diff
@@ -76,32 +83,53 @@ export function partialMemo(renderable: Renderable, existingId?: number): Node |
 
     // N.B. Async context doesn't get bound to the generator, so we need to do that manually.
     const generator = bindAsyncGeneratorToActiveContext(unboundGenerator);
-    const sink: (Renderable | typeof AppendOnlyStream)[] = [];
-    let finalResult: Renderable | typeof AppendOnlyStream = null;
+    const sink: (Node | AppendOnlyStreamValue)[] = [];
+
     let completed = false;
     let nextPromise: Promise<void> | null = null;
 
     return {
       [memoizedIdSymbol]: id,
-      async *[Symbol.asyncIterator](): AsyncGenerator<
-        Renderable | typeof AppendOnlyStream,
-        Renderable | typeof AppendOnlyStream
-      > {
+      async *[Symbol.asyncIterator](): AsyncGenerator<Node | AppendOnlyStreamValue, Node | AppendOnlyStreamValue> {
         let index = 0;
+        let isAppendOnly = false;
+
         while (true) {
           if (index < sink.length) {
-            yield sink[index++];
+            // There's something we can yield/return right away.
+            let nodes = [] as Node[];
+            while (index < sink.length) {
+              let value = sink[index++];
+              if (isAppendOnlyStreamValue(value)) {
+                isAppendOnly = true;
+                value = valueToAppend(value);
+              }
+
+              if (!isAppendOnly) {
+                nodes = [value];
+              } else {
+                nodes.push(value);
+              }
+            }
+
+            const valueToYield = isAppendOnly ? AppendOnlyStream(nodes) : nodes;
```

Comment:
I'm sure I'm missing something, but why yield `nodes` as opposed to `_.last(sink)`?

Reply:
Good suggestion -- refactored a bit to make the loop more explicitly only in service to the `isAppendOnly` scenario.